### PR TITLE
Fix #9: Add support for differing BitsPerPixel and Depth settings

### LINKED ISSUE #9: Force Color Display
I am having trouble trying to figure out how to change its color display. So if it connects to a viewer I can tell it to connect in 8 bit or 16 bit. I have tried adjusting the RGB but it really did not work. 

Any help would be great to achomplish this. I want to use this viewer over low bandwidth.

### PR DESCRIPTION
**Pull request by zzincubus**
Fix #9: Add support for differing BitsPerPixel and Depth settings

### CODE REVIEW

File: VncSharp/Framebuffer.cs [hunk 1]
```diff
@@ -79,7 +79,7 @@ public Rectangle Rectangle
 	    }
 
 	    /// <summary>
-		/// The number of Bits Per Pixel for the Framebuffer--one of 8, 24, or 32.
+		/// The number of Bits Per Pixel for the Framebuffer--one of 8, 16, or 32.
```
Comment (humphd) [inline VncSharp/Framebuffer.cs]: What happened to 24?

File: VncSharp/Framebuffer.cs [hunk 2]
```diff
@@ -168,36 +168,104 @@ public byte[] ToPixelFormat()
...
-		        GreenShift = b[11],
-		        BlueShift = b[12]
-		    };
-
-		    // Last 3 bytes are padding, ignore									
-
-			return buffer;
-		}
+
+        /// <summary>
```
Comment (humphd) [inline VncSharp/Framebuffer.cs]: Can you fix the indenting on this block?  It seems to have all gotten shifted.

File: VncSharp/RemoteDesktop.cs [hunk 3]
```diff
@@ -634,8 +658,9 @@ protected override void OnPaint(PaintEventArgs pe)
                 switch (state)
                 {
                     case RuntimeState.Connected:
-                        Assert(desktop != null);
-                        DrawDesktopImage(desktop, pe.Graphics);
+                        // Assert(desktop != null);
```
Comment (humphd) [inline VncSharp/RemoteDesktop.cs]: Just remove this completely.

File: VncSharp/RemoteDesktop.cs [hunk 4]
```diff
@@ -651,8 +676,9 @@ protected override void OnPaint(PaintEventArgs pe)
             else
             {
                 // Draw a static screenshot of a Windows desktop to simulate the control in action
-                Assert(designModeDesktop != null);
-                DrawDesktopImage(designModeDesktop, pe.Graphics);
+                // Assert(designModeDesktop != null);
```
Comment (humphd) [inline VncSharp/RemoteDesktop.cs]: Same thing here, just drop it.

File: VncSharp/RfbProtocol.cs [hunk 5]
```diff
@@ -186,14 +186,36 @@ public void Close()
 		public void ReadProtocolVersion()
 		{
 			var b = Reader.ReadBytes(12);
-
-			// As of the time of writing, the only supported versions are 3.3, 3.7, and 3.8.
-			if (Encoding.ASCII.GetString(b) == RFB_VERSION_ZERO) // Repeater functionality
+            string a = "";
```
Comment (humphd) [inline VncSharp/RfbProtocol.cs]: Indent issue here again.

File: VncSharp/VncClient.cs [hunk 6]
```diff
@@ -318,7 +319,8 @@ public void StartUpdates()
 		public void Disconnect()
 		{
 			// Stop the worker thread.
-			done.Set();
+			if (done != null)
+                done.Set();
```
Comment (humphd) [inline VncSharp/VncClient.cs]: Indent issue here

File: VncSharp/VncClient.cs [hunk 7]
```diff
@@ -328,8 +330,8 @@ public void Disconnect()
 				// this may not work, as Disconnect can get called in response to the
 				// VncClient raising a ConnectionLost event (e.g., the remote host died).
 			}
-
-			worker.Join(3000);	// this number is arbitrary, just so that it doesn't block forever....
+            if (worker != null)
```
Comment (humphd) [inline VncSharp/VncClient.cs]: Same here.

File: VncSharp/Framebuffer.cs (hunk 1 above)
Comment (zzincubus) [reply VncSharp/Framebuffer.cs]: As per the rfbproto.h document from http://www.hep.phy.cam.ac.uk/vnc_docs/, there is only bitsPerPixel values of 8, 16, or 32 (extract shown below).

/*−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−−
* Structure used to specify pixel format.
*/
typedef struct {
CARD8 bitsPerPixel; /* 8,16,32 only */

Comment (humphd) [reply VncSharp/Framebuffer.cs]: Indeed you're right.  https://tools.ietf.org/html/rfc6143#section-7.4 also agrees with this.

**Comment by zzincubus** [comment]
I have made further changes to the network code. This allows better monitoring using Wireshark. Also made changes that improve support for slow network connections. 
Moved where the RequestScreenUpdate  function is called, to prevent it being multiple times (when more than one rectangle update has been returned by the server). Have tested over slow network connections, need to stress test to ensure this is the correct thing to do.
The multiple screen updates were seen while using Wireshark.

**Comment by humphd** [comment]
@zzincubus if you want to make different changes, can you do them in separate branches/pull requests?  It makes it easier to roll things back if we only change 1 thing per merge.  If changes are related, obviously keep them together; otherwise, let's keep them isolated.

In terms of git/github workflow, do this for each different thing:

```
# make sure you have my repo added as a remote called `upstream` so you can pull changes
# should only need to do this once
git remote add upstream https://github.com/humphd/VncSharp.git

# now make sure your `master` branch is up-to-date with any new stuff.
# do this everytime before you create a new branch, so you're in sync with upstream
git checkout master
git pull upstream master

# now make a new branch and do your work
git checkout -b name-of-branch
...work...
git add <file> <file>
git commit -m "Fix XYZ"
git push origin name-of-branch

# make a new pull request
```

I'd love for someone to review/test your work (I don't have a windows box atm), but if no one is going to step up, I'll probably just merge this.  The code looks pretty good to me.  I'll comment inline.

File: VncSharp/RemoteDesktop.cs [hunk 8]
```diff
@@ -271,7 +273,10 @@ private void VncUpdate(object sender, VncEventArgs e)
             Invalidate(desktopPolicy.AdjustUpdateRectangle(e.DesktopUpdater.UpdateRectangle));
 
             if (state != RuntimeState.Connected) return;
-            vnc.RequestScreenUpdate(fullScreenRefresh);
+            // Do not request an update here, as this can be called for multiple time
+            // for one server response (multiple rectangles returned from the host)
+            // Moved to the GetRfbUpdates function (21-02-2018)
+            //vnc.RequestScreenUpdate(fullScreenRefresh);
```
Comment (humphd) [inline VncSharp/RemoteDesktop.cs]: Just delete all this instead of leaving commented code.

File: VncSharp/RfbProtocol.cs [hunk 9]
```diff
@@ -151,11 +152,16 @@ public void Connect(string host, int port)
...
+			tcp = new TcpClient {NoDelay = true}; // turn-off Nagle's Algorithm for better interactive performance with host.
 
-		    tcp.Connect(host, port);
+			tcp.ReceiveTimeout = 15000; // set receive timeout to 15s
+			tcp.SendTimeout = 15000;    // set send timeout to 15s
+			tcp.Connect(host, port);
 			stream = tcp.GetStream();
 
+			stream.ReadTimeout = 15000; // set read timeout to 15s
+			stream.WriteTimeout = 15000;// set write timeout to 15s
```
Comment (humphd) [inline VncSharp/RfbProtocol.cs]: Define a constant above for all these timeouts so it's documented once.

File: VncSharp/RfbProtocol.cs [hunk 10]
```diff
@@ -390,15 +425,17 @@ public void WriteSetPixelFormat(Framebuffer buffer)
...
+			byte[] buff = new byte[(encodings.Length*4) + 4];
+			int x = 0;
+			buff[0] = SET_ENCODINGS;
+			buff[1] = 0x00;
+			BitConverter.GetBytes((ushort)encodings.Length).Reverse().ToArray().CopyTo(buff, 2);
 			
 			foreach (var t in encodings)
-			{
-			    writer.Write(t);
+			{	BitConverter.GetBytes(t).Reverse().ToArray().CopyTo(buff, 4+x);
```
Comment (humphd) [inline VncSharp/RfbProtocol.cs]: Newline after `{` please

File: VncSharp/RfbProtocol.cs [hunk 11]
```diff
@@ -19,6 +19,7 @@
 using System.Diagnostics;
 using System.Drawing;
 using System.IO;
+using System.Linq;
```
Comment (humphd) [inline VncSharp/RfbProtocol.cs]: What's this for? (mostly curious, fine if you need it.  I'm not clear why it's being added)

File: VncSharp/RfbProtocol.cs [hunk 12]
```diff
@@ -808,9 +857,15 @@ public void DecodeStream()
...
 							bytesToRead = maxBufferSize;
-
-						// try reading bytes
-						var bytesRead = netStream.Read(receiveBuffer, 0, bytesToRead);
+						
+						// try reading bytes (read in 1024 byte chunks) - improvement for slow connections
+						int toRead = (bytesToRead > 1024) ? 1024 : bytesToRead;
+						int bytesRead = 0;
+						try
+						{	bytesRead = netStream.Read(receiveBuffer, bytesRead, toRead);
```
Comment (humphd) [inline VncSharp/RfbProtocol.cs]: Newline after `{` please

File: VncSharp/RfbProtocol.cs (hunk 11 above)
Comment (zzincubus) [reply VncSharp/RfbProtocol.cs]: This is used for the following (Reverse):
BitConverter.GetBytes((ushort)x).Reverse().ToArray().CopyTo(buff, 2);

Comment (humphd) [reply VncSharp/RfbProtocol.cs]: Thanks.  .NET has evolved a lot since I wrote this code, so I wondered.

**Comment by zzincubus** [comment]
I think the above are related due to the differing bits / depth allowing for slower connections. These changes also improve connections on slower (GSM) connections.

File: VncSharp/RfbProtocol.cs (hunk 11 above)
Comment (zzincubus) [reply VncSharp/RfbProtocol.cs]: No problem :-)

**Comment by humphd** [comment]
@Fabzien or @gashlin any chance either of you would be willing to test this?  You can download @zzincubus's work at https://github.com/zzincubus/VncSharp/archive/force-color-display.zip.

@zzincubus do you have more changes you want to do here, or you feeling done?  I think the code is fine, but I'd love to get someone to test before I merge, and I don't have a Windows machine right now.  If no one steps up soon, and you're happy with this, I'll probably just land it and we can fix any regressions in later PRs.

**Comment by zzincubus** [comment]
I have no further changes to add to this. If there are any issues with the changes that have been made, I will try to resolve as and when they are encountered. I have been happy with the testing I have performed on this (monitoring PLCs and various Linux machines).

**Comment by humphd** [comment]
@zzincubus forgive me if this is obvious, but to get your changes on `master` do:

```
git checkout master
git pull upstream master
```

And then if you ever want to fix something else, `git checkout -b new-fix` and do your work there.  Thanks again.

**Comment by zzincubus** [comment]
OK thank you for the instructions, and thanks for accepting the changes.

**Pull request merged**